src/rtctools_heat_network/workflows/grow_workflow.py

```diff
@@ -523,18 +539,58 @@ def run_end_scenario_sizing(end_scenario_problem_class, staged_pipe_optimization
     if staged_pipe_optimization:
         solution = run_optimization_problem(end_scenario_problem_class, stage=1)
         results = solution.extract_results()
+        parameters = solution.parameters(0)
 
         # We give bounds for stage 2 by allowing one DN sizes larger than what was found in the
         # stage 1 optimization.
         pc_map = solution._HeatMixin__pipe_topo_pipe_class_map
         for pipe_classes in pc_map.values():
             v_prev = 0.0
+            first_pipe_class = True
             for var_name in pipe_classes.values():
                 v = results[var_name][0]
-                boolean_bounds[var_name] = (0.0, abs(v))
-                if v_prev == 1.0:
+                if first_pipe_class and abs(v) == 1.0:
+                    boolean_bounds[var_name] = (abs(v), abs(v))
+                elif abs(v) == 1.0:
+                    boolean_bounds[var_name] = (0.0, abs(v))
+                elif v_prev == 1.0:
                     boolean_bounds[var_name] = (0.0, 1.0)
+                else:
+                    boolean_bounds[var_name] = (abs(v), abs(v))
                 v_prev = v
+                first_pipe_class = False
+
+        for asset in [
+            *solution.heat_network_components.get("source", []),
+            *solution.heat_network_components.get("buffer", []),
+        ]:
+            var_name = f"{asset}_aggregation_count"
+            r = results[var_name][0]
+            if round(r) >= 1:
+                boolean_bounds[var_name] = (r, r)
+
+        t = solution.times()
+        from rtctools.optimization.timeseries import Timeseries
+
+        for p in solution.heat_network_components.get("pipe", []):
+            if p in solution.hot_pipes and parameters[f"{p}.area"] > 0.0:
+                lb = []
+                ub = []
+                for i in range(len(t)):
+                    r = results[f"{p}__flow_direct_var"][i]
+                    lb.append(
+                        r if abs(results[f"{p}.Q"][i] / parameters[f"{p}.area"]) > 1.0e-2 else 0
+                    )
+                    ub.append(
+                        r if abs(results[f"{p}.Q"][i] / parameters[f"{p}.area"]) > 1.0e-2 else 1
+                    )
```

Comment:
This condition is a bit strict I believe. It is now stated that the flow direction is fixed if the velocity is larger than 0.02m/s, I think we should increase the velocity limit to atleast 0.1 or 0.5 m/s because the dimension of the pipe is already optimised (might only go 1 up). The flow in some pipes, particularly the smaller ones might switch. The current example will most likely not run in this problem because of the tree structure with mainly the central producers, but other networks might.

Reply:
this is a bit of a judgement call, I checked this as being larger then the flow needed to compensate heat losses. We can up it a bit, but not to 0.1m/s I would say is already on the high side

Reply:
You are missing the point. It is not about being larger than the heat losses, it was just to have an indication of what is a reasonable minimum flow to determine direction on. This is a potential situation which could be better, whereas the point of this staged approach is to guide stage 2 and steer it to a solution we know is good (maybe not optimal), therefore it might be possible that you can find another solution but a solution with these flow direction should also be a good solution.